### PR TITLE
Add support for shared_memory for Kubernetes workloads.

### PR DESCRIPTION
**Pull request by shrinandj**
This includes support for Kubernetes jobs as well as Argo workflows.

Testing Done:

- Verified that setting shared_memory=N in @kubernetes works as expected
- Verified that setting shared_memory=N in @kubernetes works with argo-workflows
- Verified that setting shared_memory=N on the cli works as expected
- Verified that not setting shared_memory works fine

### CODE REVIEW

**Comment by roofurmston** [comment]
Perhaps I am misunderstanding the PR, apologies if that is the case, but isn't shared memory already supported in kubernetes through the `use_tmpfs`, `tmpfs_path` and `tmpfs_size` parameters?

**Comment by saikonen** [comment]
> Perhaps I am misunderstanding the PR, apologies if that is the case, but isn't shared memory already supported in kubernetes through the `use_tmpfs`, `tmpfs_path` and `tmpfs_size` parameters?

functionally there is no difference, and it can be supported with the tmpfs parameters as you noted. This is most likely a convenience/convergence feature, as the `shared_memory` is a separate parameter on `batch` for example

**Review by saikonen — COMMENTED** [review]
this doesn't seem to work with the `resources` decorator yet? Is it in scope for the PR?

```python
@resources(shared_memory=120)
@kubernetes(shared_memory=100)
@step
```
the expectation is that the larger of the two being picked.

**Comment by roofurmston** [comment]
> > Perhaps I am misunderstanding the PR, apologies if that is the case, but isn't shared memory already supported in kubernetes through the `use_tmpfs`, `tmpfs_path` and `tmpfs_size` parameters?
> 
> functionally there is no difference, and it can be supported with the tmpfs parameters as you noted. This is most likely a convenience/convergence feature, as the `shared_memory` is a separate parameter on `batch` for example

OK, fair enough. Just want to raise it really.

**Comment by shrinandj** [comment]
Thanks for the questions/review.

One reason to have an explicit `shared_memory` option would also be if someone wants the use both: tmpfs and shared_memory. tmpfs is be more generic. `shared_memory` specifically gets used by some of the existing tools like deepspeed and MPI.

**Review by saikonen — CHANGES_REQUESTED** [review]
the stubs test cases need a fix before we go ahead with this, as its going to block the release otherwise due to failures. Not sure why they're failing with a simple change to the decorator func signature. @madhur-ob do you have a chance to take a look at the failures?

File: metaflow/plugins/kubernetes/kubernetes_cli.py [hunk 1]
```diff
@@ -107,6 +107,7 @@ def kubernetes():
     type=JSONTypeClass(),
     multiple=False,
 )
+@click.option("--shared-memory", default=None, help="Size of shared memory in MB")
```
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kubernetes_cli.py:110]: I believe it's not MB but MiB. We should look into other values like disk, tmpfs, memory and ensure that all fields have the same dimension

**Review by savingoyal — DISMISSED** [review]

File: metaflow/plugins/kubernetes/kubernetes_cli.py (hunk 1 above)
Comment (shrinandj) [reply metaflow/plugins/kubernetes/kubernetes_cli.py:110]: Fixed.

**Comment by madhur-ob** [comment]
@shrinandj I guess you reverted my changes by accident?

**Review by saikonen — APPROVED** [review]

**Pull request merged**